teps/0146-parameters-in-script.md

```diff
@@ -0,0 +1,605 @@
+values are stored in [TEP-0086][tep-0086-alt]. In that proposal a sidecar is
+responsible for fetching `Results` from previous `Tasks` and writing them to
+a location on disk for use as `Parameters`.
+
+##### Discussion
+
+This approach aligns well with the proposal in [TEP-0086][tep-0086-prop], and
+brings consistency in how we supply both `Parameters` and `Results` to `Steps`.
+
+###  Combine Elements of Both Explicit Environment Variable and Filesystem Projection Solutions
+
+Building on environment variable alternatives (1.0 & 1.1), and filsystem alternatives (2) Tekton Pipelines could offer an explicit option to the user whether to project as env (language agnostic OR shell format) or file.  There could be different env syntaxes for shell (most used in tasks) vs language agnostic env vars to make the option more UX friendly and backwards compatible for the majority of tasks (which use `sh` or `bash`).
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone $(params.url)
+  
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone ${$(params.url1.env)}
+      git clone $(params.url2.env_shell)
+      git clone $(params.url3.path)
+      
+  # After - Resolved
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_URL1
+      value: $(params.url1)
+    script: |
+      git clone ${PARAM_URL1}
+      git clone ${PARAM_URL2}
+      IFS='' read -d -r PARAM_URL3 /tekton/params/url3
+      git clone ${PARAM_URL3}
+```
+
+####  Discussion
+
+Similar to the other options this mutates the spec, though does so at the user's
+discretion. This alternative could be an extension of 1.0 and could be implemented piecemeal with only a single implicit env format supported initially (`.env`) with additional implicit env formats and filesystem (`$(params.bar.path)`) support added later if needed.
+
+###  Reject `Parameters` in `Script` and provide instructions to fix
+
+During validation we could detect `Parameters` used in `Script` blocks and return
+a helpful message to the user that they need to rewrite their `Task` to move the
+`Parameters` into environment variables.
+
+```yaml
+  # Before
+  steps:
+  - name: foo
+    image: myimage
+    script: |
+      git clone $(params.url)
+```
+
+When the Tekton Pipelines Controller sees this `Script` it rejects the `Task` submission
+with an error, such as:
+
+```
+spec.steps[0].script: $(params.url) cannot be injected into a script, consider moving
+into an environment variable on the Step.
+```
+
+```yaml
+  # After
+  steps:
+  - name: foo
+    image: myimage
+    env:
+    - name: PARAM_URL
+      value: $(params.url)
+    script: |
+      git clone "${PARAM_URL}"
+```
+
+##### Discussion
+
+Non-mutating but doesn't help the user as directly as mutating the spec.
+
+
+## Implementation Plan
+
+1. Introduce the `params.<name>.env` syntax and document the new approach.
```

Comment:
Does this assume that one of the alternatives has been chosen?
What does the `params.<name>.env` refers to, is this the "Combine Elements of Both Explicit Environment Variable and Filesystem Projection Solutions" alternative?

Reply:
Sorry for the confusion here, the “params.<name>.env” here is referring to the preferred alternative from the document which is “Suggested Solution - Explicitly Project `Parameters` as Environment Variables in `Steps` (Language-Agnostic)”.  I have updated the snippet here to now reference that preferred solution by name.  “Combine Elements of Both Explicit Environment Variable and Filesystem Projection Solution” is a possible extension point once “Explicitly Project `Parameters` as Environment Variables in `Steps` (Language-Agnostic)” is implemented. That way we can start and then extend as needed for UX, additional options to aid in more automated migration, etc.